examples/custom_storage.ml

```diff
@@ -0,0 +1,105 @@
+(*
+ * Copyright (c) 2022 Tarides <contact@tarides.com>
+ *
+ * Permission to use, copy, modify, and distribute this software for any
+ * purpose with or without fee is hereby granted, provided that the above
+ * copyright notice and this permission notice appear in all copies.
+ *
+ * THE SOFTWARE IS PROVIDED "AS IS" AND THE AUTHOR DISCLAIMS ALL WARRANTIES
+ * WITH REGARD TO THIS SOFTWARE INCLUDING ALL IMPLIED WARRANTIES OF
+ * MERCHANTABILITY AND FITNESS. IN NO EVENT SHALL THE AUTHOR BE LIABLE FOR
+ * ANY SPECIAL, DIRECT, INDIRECT, OR CONSEQUENTIAL DAMAGES OR ANY DAMAGES
+ * WHATSOEVER RESULTING FROM LOSS OF USE, DATA OR PROFITS, WHETHER IN AN
+ * ACTION OF CONTRACT, NEGLIGENCE OR OTHER TORTIOUS ACTION, ARISING OUT OF
+ * OR IN CONNECTION WITH THE USE OR PERFORMANCE OF THIS SOFTWARE.
+ *)
+
+open Lwt.Syntax
+
+(** Create a configuration module for our storage. Here we demonstrate a simple
+    configuration for setting the initial size of the hash table. *)
+
+module Hashtbl_config = struct
+  module Conf = Irmin.Backend.Conf
+
+  let spec = Conf.Spec.v "hashtbl"
+  let init_size = Conf.key ~spec "init-size" Irmin.Type.int 8
+  let empty = Conf.empty spec
+end
+
+(** Create a {!Irmin.Storage.Make} functor for our hash table storage. *)
+
+module Hashtbl_storage : Irmin.Storage.Make =
+functor
+  (Key : Irmin.Type.S)
+  (Value : Irmin.Type.S)
+  ->
+  struct
+    module Tbl = Hashtbl.Make (struct
+      type t = Key.t
+
+      let equal a b = Irmin.Type.(unstage (equal Key.t)) a b
+      let hash k = Irmin.Type.(unstage (short_hash Key.t)) k
+    end)
+
+    (** Types *)
+
+    type 'a t = { t : Value.t Tbl.t; l : Mutex.t }
+    (** NOTE: the 'a type parameter is used by Irmin for tracking store
```

Comment:
Maybe we do not want to expose this in the "simple" storage backend - (and also emove the cast function)

Reply:
Yeah, I think this is a good call.